configure: error: Could not link against boost_system-mt !

When executing:
`./zcutil/build.sh --disable-rust -j$(nproc)`

I get:
```
...
checking miniupnpc/upnperrors.h presence... yes
checking for miniupnpc/upnperrors.h... yes
checking for main in -lminiupnpc... (cached) yes
checking for boostlib >= 1.20.0... yes
checking whether the Boost::System library is available... yes
checking for exit in -lboost_system-mt... no
configure: error: Could not link against boost_system-mt !
```

### The version of Zcash you were using:
v1.0.8-1

### Machine specs:
```
uname -or
4.10.0-20-generic GNU/Linux
uname -a
Linux wks-PC 4.10.0-20-generic #22-Ubuntu SMP Thu Apr 20 09:22:42 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

 gcc --version
gcc (Ubuntu 7-20170407-0ubuntu2) 7.0.1 20170407 (experimental) [trunk revision 246759]
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

same on gentoo

`make: opzione non riconosciuta --disable-proton`

and compiling zcash after fetch-params.sh results in this
`Building boost...
mkdir -p /home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/.
cd /home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/.; PATH=/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/x86_64-unknown-linux-gnu/native/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/x86_64-pc-linux-gnu/gcc-bin/4.9.4:/usr/lib/llvm/5/bin:/usr/lib/llvm/4/bin:/usr/games/bin   ./b2 -d2 -j2 -d1 --prefix=/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/staging/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/x86_64-unknown-linux-gnu --layout=tagged --build-type=complete --user-config=user-config.jam threading=multi link=static -sNO_BZIP2=1 -sNO_ZLIB=1 variant=release   threadapi=pthread runtime-link=shared    stage
/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/tools/build/src/build/feature.jam:494: in feature.validate-value-string from module feature
error: "none" is not a known value of feature <optimization>
error: legal values: "off" "speed" "space"
/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/tools/build/src/build/property.jam:276: in validate1 from module property
/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/tools/build/src/build/property.jam:302: in property.validate from module property
/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/tools/build/src/tools/builtin.jam:381: in variant from module builtin
/usr/share/boost-build/site-config.jam:9: in modules.load from module site-config
/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/tools/build/src/build-system.jam:249: in load-config from module build-system
/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/tools/build/src/build-system.jam:351: in load-configuration-files from module build-system
/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/tools/build/src/build-system.jam:524: in load from module build-system
/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/tools/build/src/kernel/modules.jam:295: in import from module modules
/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/tools/build/src/kernel/bootstrap.jam:139: in boost-build from module
/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/boost-build.jam:17: in module scope from module

make: *** [funcs.mk:240: /home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends/work/build/x86_64-unknown-linux-gnu/boost/1_62_0-2d69d505974/./.stamp_built] Error 1
make: uscita dalla directory "/home/user/work/dude/activity/z_c_ash_wa_llet/zcash/depends"

`

The same issue here, I'm not sure, but it has something to do with g++, which I have in version 7.1.1 - it doesn't allow to override gcc internal prototype - it seems to be ok on older versions of gcc, I think.
`conftest.cpp:71:6: error: declaration of 'char exit()' conflicts with built-in declaration 'void exit(int)' [-Werror=builtin-declaration-mismatch]`

I'm seeing the same issue as OP here on Arch Linux against tagged release v1.0.10-1 with gcc 7.1.1 and boost 1.64.0-3. The config.log also shows the same issue reported by @vrfpk which he was told was a separate issue, but I'm not convinced it is.
```
checking for exit in -lboost_system-mt... no
configure: error: Could not link against boost_system-mt !
```

Here are some relevant lines from config.log
```
configure:21996: checking for boostlib >= 1.20.0                                                                                                                                                                                                                                                                             
configure:22081: g++ -m64 -c -std=c++11 -pipe -O1 -fwrapv -fno-strict-aliasing -Werror -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE -I/home/ayleph/zcash/depends/x86_64-unknown-linux-gnu/share/../include/  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I/home/ayleph/zcash/depends/x86_64-unknown-linux-gnu/share/../include conftest.cpp >&5                                                                                                                                                                                                                                      
configure:22081: $? = 0                                                                                                                                                                                                                                                                                                      
configure:22083: result: yes                                                                                                                                                                                                                                                                                                 
configure:22267: checking whether the Boost::System library is available                                                                                                                                                                                                                                                     
configure:22291: g++ -m64 -c -std=c++11 -pipe -O1 -fwrapv -fno-strict-aliasing -Werror -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE -I/home/ayleph/zcash/depends/x86_64-unknown-linux-gnu/share/../include/  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I/home/ayleph/zcash/depends/x86_64-unknown-linux-gnu/share/../include conftest.cpp >&5                                                                                                                                                                                                                                      
configure:22291: $? = 0                                                                                                                                                                                                                                                                                                      
configure:22306: result: yes                                                                                                                                                                                                                                                                                                 
configure:22322: checking for exit in -lboost_system-mt                                                                                                                                                                                                                                                                      
configure:22347: g++ -m64 -o conftest -std=c++11 -pipe -O1 -fwrapv -fno-strict-aliasing -Werror -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE -I/home/ayleph/zcash/depends/x86_64-unknown-linux-gnu/share/../include/  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I/home/ayleph/zcash/depends/x86_64-unknown-linux-gnu/share/../include -L/home/ayleph/zcash/depends/x86_64-unknown-linux-gnu/share/../lib   -Wl,-z,relro -Wl,-z,now -pie -L/home/ayleph/zcash/depends/x86_64-unknown-linux-gnu/share/../lib conftest.cpp -lboost_system-mt  -lanl  >&5                              
conftest.cpp:71:6: error: declaration of 'char exit()' conflicts with built-in declaration 'void exit(int)' [-Werror=builtin-declaration-mismatch]                                                                                                                                                                           
 char exit ();                                                                                                                                                                                                                                                                                                               
      ^~~~ 
cc1plus: all warnings being treated as errors                                                                                                                                                                                                                                                                                
configure:22347: $? = 1                                                                                                                                                                                                                                                                                                      
configure: failed program was:                                                                                                                                                                                                                                                                                               
| /* confdefs.h */                                                                                                                                                                                                                                                                                                           <snip>                                                                                                                                                                                                                                                                                
| /* end confdefs.h.  */
|                                                                                                                                                                                                                                                                                                                            
| /* Override any GCC internal prototype to avoid an error.                                                                                                                                                                                                                                                                  
|    Use char because int might match the return type of a GCC                                                                                                                                                                                                                                                               
|    builtin and then its argument prototype would still apply.  */                                                                                                                                                                                                                                                          
| #ifdef __cplusplus                                                                                                                                                                                                                                                                                                         
| extern "C"                                                                                                                                                                                                                                                                                                                 
| #endif                                                                                                                                                                                                                                                                                                                     
| char exit ();                                                                                                                                                                                                                                                                                                              
| int                                                                                                                                                                                                                                                                                                                        
| main ()                                                                                                                                                                                                                                                                                                                    
| {                                                                                                                                                                                                                                                                                                                          
| return exit ();                                                                                                                                                                                                                                                                                                            
|   ;                                                                                                                                                                                                                                                                                                                        
|   return 0;                                                                                                                                                                                                                                                                                                                
| }                                                                                                                                                                                                                                                                                                                          
configure:22357: result: no                                                                                                                                                                                                                                                                                                  
configure:22469: error: Could not link against boost_system-mt !
```

Hi, I'm having the same problem in arch linux:

`checking for exit in -lboost_system-mt... no`
`configure: error: Could not link against boost_system-mt !`

I'm on the latest commit where @radix42 applied the fix for arch.
I'd also checked gcc5, boost and boost-libs are installed.

I tried a lot of possible fixes but none works.

Thanks.

A workaround is to remove `-Werror` from the ./configure command in `zcutil/build.sh`

The problem can be seen in config.log and is casused by a warning with newer compilers which with that compiler switch is treated as an error. 

From config.log:
```
configure:22242: checking for exit in -lboost_system-mt
configure:22267: g++ -m64 -o conftest -std=c++11 -pipe -fPIC -O1 -fwrapv -fno-strict-aliasing -Werror -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE ..<snip>... /x86_64-unknown-linux-gnu/share/../lib conftest.cpp -lboost_system-mt  -lanl  >&5
conftest.cpp:70:6: error: declaration of 'char exit()' conflicts with built-in declaration 'void exit(int)' [-Werror=builtin-declaration-mismatch]
 char exit ();
      ^~~~
cc1plus: all warnings being treated as errors

```

Nice, it works on manjaro (arch based distro). Thanks!

On Thu, Jul 20, 2017 at 12:00 PM, Bjorn Hjortsberg <notifications@github.com
> wrote:

> A workaround is to remove -Werror from the ./configure command in
> zcutil/build.sh
>
> The problem can be seen in config.log and is casused by a warning with
> newer compilers which with that compiler switch is treated as an error.
>
> From config.log:
>
> configure:22242: checking for exit in -lboost_system-mt
> configure:22267: g++ -m64 -o conftest -std=c++11 -pipe -fPIC -O1 -fwrapv -fno-strict-aliasing -Werror -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE ..<snip>... /x86_64-unknown-linux-gnu/share/../lib conftest.cpp -lboost_system-mt  -lanl  >&5
> conftest.cpp:70:6: error: declaration of 'char exit()' conflicts with built-in declaration 'void exit(int)' [-Werror=builtin-declaration-mismatch]
>  char exit ();
>       ^~~~
> cc1plus: all warnings being treated as errors
>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/zcash/zcash/issues/2304#issuecomment-316657338>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEhGyNhK_wZTu0IatzKFJom1RUC_38zjks5sPyVJgaJpZM4NKGlk>
> .
>



-- 
Pozdrav,
Nikola

--- a/configure.ac
+++ b/configure.ac
@@ -720,7 +720,7 @@ AX_CHECK_COMPILE_FLAG([-fwrapv],[CXXFLAGS="$CXXFLAGS -fwrapv"])
 AX_CHECK_COMPILE_FLAG([-fno-strict-aliasing],[CXXFLAGS="$CXXFLAGS -fno-strict-aliasing"])
 AX_CHECK_COMPILE_FLAG([-Wno-builtin-declaration-mismatch],[CXXFLAGS="$CXXFLAGS -Wno-builtin-declaration-mismatch"],,[[$CXXFLAG_WERROR]])
 
-LIBZCASH_LIBS="-lgmp -lgmpxx -lboost_system-mt -lcrypto -lsodium $RUST_LIBS"
+LIBZCASH_LIBS="-lgmp -lgmpxx -lboost_system -lcrypto -lsodium $RUST_LIBS"
 
 AC_MSG_CHECKING([whether to build bitcoind])
 AM_CONDITIONAL([BUILD_BITCOIND], [test x$build_bitcoind = xyes])

--- a/depends/packages/boost.mk
+++ b/depends/packages/boost.mk
@@ -9,7 +9,7 @@ define $(package)_set_vars
 $(package)_config_opts_release=variant=release
 $(package)_config_opts_debug=variant=debug
 $(package)_config_opts=--layout=tagged --build-type=complete --user-config=user-config.jam
-$(package)_config_opts+=threading=multi link=static -sNO_BZIP2=1 -sNO_ZLIB=1
+$(package)_config_opts+=link=static -sNO_BZIP2=1 -sNO_ZLIB=1
 $(package)_config_opts_linux=threadapi=pthread runtime-link=shared
 $(package)_config_opts_darwin=--toolset=darwin-4.2.1 runtime-link=shared
 $(package)_config_opts_mingw32=binary-format=pe target-os=windows threadapi=win32 runtime-link=static

